File: kotlin/meetup/src/main/kotlin/Meetup.kt
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.YearMonth

class Meetup(month: Int, year: Int) {

    private val yearMonth = YearMonth.of(year, month)

    fun day(dayOfWeek: DayOfWeek, schedule: MeetupSchedule): LocalDate {
        var dayOffset = 1 + getWeekOffset(schedule) * 7

        val firstDay = yearMonth.atDay(1).dayOfWeek

        if (firstDay > dayOfWeek) {
            dayOffset += 7 - firstDay.value
        }


        val d = yearMonth.atDay(dayOffset)
        return d
    }

    private fun getWeekOffset(schedule: MeetupSchedule): Int {
        return when (schedule) {
            MeetupSchedule.FIRST -> 0
            MeetupSchedule.SECOND -> 1
            MeetupSchedule.THIRD -> 2
            MeetupSchedule.FOURTH -> 3
            MeetupSchedule.LAST -> TODO()
            MeetupSchedule.TEENTH -> TODO()
        }
    }
}

File: go/meetup/meetup.go
package meetup

import "time"

type WeekSchedule int

const (
	First WeekSchedule = iota
	Second
	Third
	Fourth
	Teenth
	Last
)

const week = 7

func Day(wSched WeekSchedule, wDay time.Weekday, month time.Month, year int) int {
	switch wSched {
	case First, Second, Third, Fourth:
		day := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
		day = day.AddDate(0, 0, week*int(wSched))
		daysToAdd := toAdd(day, wDay)
		return day.AddDate(0, 0, daysToAdd).Day()
	case Teenth:
		day := time.Date(year, month, 13, 0, 0, 0, 0, time.UTC)
		daysToAdd := toAdd(day, wDay)
		return day.AddDate(0, 0, daysToAdd).Day()
	case Last:
		day := time.Date(year, month+1, 1, 0, 0, 0, 0, time.UTC).Add(-1 * time.Minute)
		daysToSub := toSub(day, wDay)
		return day.AddDate(0, 0, -daysToSub).Day()
	}
	return 0
}

func toAdd(current time.Time, wDay time.Weekday) int {
	cDay := current.Weekday()
	if cDay > wDay {
		return week + int(wDay-cDay)
	}
	return int(wDay - cDay)
}

func toSub(current time.Time, wDay time.Weekday) int {
	cDay := current.Weekday()
	if cDay < wDay {
		return week - int(wDay-cDay)
	}
	return int(cDay - wDay)
}
